pkg/filters/proxy/loadbalance.go

```diff
@@ -38,53 +41,144 @@ const (
 	LoadBalancePolicyIPHash = "ipHash"
 	// LoadBalancePolicyHeaderHash is the load balance policy of HTTP header hash.
 	LoadBalancePolicyHeaderHash = "headerHash"
+	// LoadBalanceStickyCookie is the generated cookie name used for sticky session
+	LoadBalanceStickyCookie = "EASEGRESS_SESSION"
 )
 
 // LoadBalancer is the interface of an HTTP load balancer.
 type LoadBalancer interface {
 	ChooseServer(req *httpprot.Request) *Server
+	Manipulate(server *Server, req *httpprot.Request, resp *httpprot.Response)
+	GetSlots() []*Server
 }
 
 // LoadBalanceSpec is the spec to create a load balancer.
 type LoadBalanceSpec struct {
 	Policy        string `json:"policy" jsonschema:"omitempty,enum=,enum=roundRobin,enum=random,enum=weightedRandom,enum=ipHash,enum=headerHash"`
 	HeaderHashKey string `json:"headerHashKey" jsonschema:"omitempty"`
+	Sticky        bool   `json:"sticky" jsonschema:"omitempty"`
+	StickyCookie  string `json:"stickyCookie" jsonschema:"omitempty"`
+	StickyExpire  string `json:"stickyExpire" jsonschema:"omitempty,format=duration"`
 }
 
 // NewLoadBalancer creates a load balancer for servers according to spec.
-func NewLoadBalancer(spec *LoadBalanceSpec, servers []*Server) LoadBalancer {
+func NewLoadBalancer(spec *LoadBalanceSpec, servers []*Server, old LoadBalancer) LoadBalancer {
+	base := BaseLoadBalancer{
+		Servers: servers,
+	}
+	var oldSlots []*Server
+	if old != nil {
+		oldSlots = old.GetSlots()
+	}
+	base.Slots = createSlots(oldSlots, servers)
+	base.confSticky(spec)
+
 	switch spec.Policy {
 	case LoadBalancePolicyRoundRobin, "":
-		return newRoundRobinLoadBalancer(servers)
+		return newRoundRobinLoadBalancer(base)
 	case LoadBalancePolicyRandom:
-		return newRandomLoadBalancer(servers)
+		return newRandomLoadBalancer(base)
 	case LoadBalancePolicyWeightedRandom:
-		return newWeightedRandomLoadBalancer(servers)
+		return newWeightedRandomLoadBalancer(base)
 	case LoadBalancePolicyIPHash:
-		return newIPHashLoadBalancer(servers)
+		return newIPHashLoadBalancer(base)
 	case LoadBalancePolicyHeaderHash:
-		return newHeaderHashLoadBalancer(servers, spec.HeaderHashKey)
+		return newHeaderHashLoadBalancer(base, spec.HeaderHashKey)
 	default:
 		logger.Errorf("unsupported load balancing policy: %s", spec.Policy)
-		return newRoundRobinLoadBalancer(servers)
+		return newRoundRobinLoadBalancer(base)
+	}
+}
+
+// confSticky configures sticky settings
+func (base *BaseLoadBalancer) confSticky(spec *LoadBalanceSpec) {
+	if spec.Sticky {
+		base.Sticky = spec.Sticky
+
+		if spec.StickyCookie == LoadBalanceStickyCookie {
+			logger.Errorf("can't use system cookie name: %s, will be ignored", LoadBalanceStickyCookie)
+			spec.StickyCookie = ""
+		}
```

Comment:
Question: does here mean if the user uses the cookie name as the same as the system name, we just use the system default cookie name? and what's the difference behavior between the system cookie and the user-defined cookie? if they have the same behavior, I think we should not log any error here.